Full Sync updates to allow full enqueuing of chunks.

mdbitz opened the pull request:
Testing of the Term_Relationships module revealed that only 10 chunks were being enqueued and processed before a jetpack_full_sync_end action was being triggered. 

#### Changes proposed in this Pull Request:
This PR modifies the Term_Relationships module to allow for enqueuing of items to match the expected value of the arguments.

In addition it modifies the continue_enqueue function in Full_Sync module so that it returns when more items exist to be processed instead of sending an early `jetpack_full_sync_end` action.

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
Bug Fix

#### Testing instructions:

* Initiate a Full Sync of the `term_relationships` module.
* `wp jetpack-sync start --blog-id=XXX --object-type=term_relationship`
* Check the Jetpack Debug page and you will see that only 10 chunks are enqueued.
* Wait for the Sync to complete
* Only 10 chunks in total will have been sent and processed.

* Apply the Patch
* Initiate a Full Sync of the `term_relationships` module.
* `wp jetpack-sync start --blog-id=XXX --object-type=term_relationship`
* Check the Jetpack Debug page and you will see that more than 10 chunks are enqueued.
* Wait for the Sync to complete
* All chunks will have been processed.

#### Proposed changelog entry for your changes:
Not sure any entry is required as this goes hand in hand with the refactoring in 13815

#### Additional Notes
Should there be additional unit test on both these functions to ensure that the max param is being adhered to and that continue_enqueue is existing correctly?

mdbitz committed:
Update continue_enqueue so that the conditional fires and returns when more enqueue actions are needed.
update check in Term_Relationships enqueue_full_sync_actions so that the limit is bases on allowing for more than 10 chunks per execution.

gravityrail commented:
I'm going to try to add some tests for this

lezama committed:
check if we have finished enqueuing the current module

lezama committed:
or maybe like this?

lezama commented on packages/sync/src/modules/Term_Relationships.php:
```diff
@@ -126,7 +126,7 @@ public function enqueue_full_sync_actions( $config, $max_items_to_enqueue, $last
 			$last_object_enqueued  = $this->bulk_enqueue_full_sync_term_relationships( $items, $last_object_enqueued );
 			$items_enqueued_count += count( $items );
-			$limit                 = min( $limit - $objects_count, self::QUERY_LIMIT );
+			$limit                 = min( $max_items_to_enqueue - $items_enqueued_count, self::QUERY_LIMIT );
```
'limit' needs to be a number of rows, each item has  `$term_relationships_full_sync_item_size`  rows. we can't use it as the limit for the query, I think.

mdbitz replied:
the issue with the existing line of code

`$limit                 = min( $limit - $objects_count, self::QUERY_LIMIT );`

is that $limit is originally set to at 106 by the min statement of
`$limit = min( $max_items_to_enqueue * $term_relationships_full_sync_item_size, self::QUERY_LIMIT );`

So when we reach this point of the code $limit has already been minified to the QUERY_LIMIT so by using $limit in or test its already been truncated to the size of a single query and we only every get 1 execution of the while statement instead of iterating to match the available space in the queue.


Perhaps it would make sense to re-use the original min statement simply removing the current limit from the first value?

`$limit = min( ($max_items_to_enqueue * $term_relationships_full_sync_item_size - $limit), self::QUERY_LIMIT );`

mdbitz replied:
scratch that it wouldn't increment in the above as we need to incrementally decrease.

mdbitz replied:
`$limit = min( ($max_items_to_enqueue  - $items_enqueued_count) * $term_relationships_full_sync_item_size, self::QUERY_LIMIT );`

So by the above if we take how much we want to enqueue subtract the number of chunks enqueued. and then times it by chunk size that gives us remainder

lezama committed:
oops

mdbitz committed:
Updated limit query to utilize chunk sizes and remaining chunks to enqueue

mdbitz commented:
Thanks Miguel, Dan and everyone for reviewing and correcting my erroneous assumptions. I like the approach of verifying all modules are complete and sorry I took the issue with Term Relationships not fully enqueuing to mean an issue with continue_enqueue.

lezama committed:
get_enqueue_status to public

so we can test it

lezama committed:
revert limit calculation

lezama committed:
re check if we have enqued all rows

We need to do this extra check in case $max_items_to_enqueue * $term_relationships_full_sync_item_size == relationships objects left.

lezama committed:
test_full_sync_enqueue_term_relationships

lezama commented on tests/php/sync/test_class.jetpack-sync-full.php:
```diff
@@ -307,6 +307,64 @@ function test_full_sync_sends_all_term_relationships() {
+
+		// how many items are we allowed to enqueue on a single request/continue_enqueuing
+		$max_enqueue_full_sync = 2;
+		// how many items are we allowed to enqueue on any given time
```
```suggestion
		// how many sync items the full sync queue can contain
```

gravityrail commented on tests/php/sync/test_class.jetpack-sync-full.php:
```diff
@@ -307,6 +307,64 @@ function test_full_sync_sends_all_term_relationships() {
+
+		$this->assertEquals( $total_items, $total );
+		$this->assertEquals( $total_items, $queued );
+		$this->assertTrue( $finished === true );
```
```suggestion
		$this->assertTrue( $finished );
```

Or

```suggestion
		$this->assertSame( $finished, true );
```

gravityrail commented on tests/php/sync/test_class.jetpack-sync-full.php:
```diff
@@ -307,6 +307,64 @@ function test_full_sync_sends_all_term_relationships() {
+
+		// hit $max_queue_size_full_sync limit
+		$status = $this->full_sync->get_enqueue_status();
+		list( $total, $queued, $finished ) = $status['term_relationships'];
```
Let's add a check that we're not finished:

```suggestion
		list( $total, $queued, $finished ) = $status['term_relationships'];
		$this->assertNotTrue( $finished );
```

gravityrail reviewed (dismissed):
LGTM, I left a couple of suggestions for tests but they're minor.

gravityrail committed:
Update tests/php/sync/test_class.jetpack-sync-full.php

Co-Authored-By: Miguel Lezama <miguel@automattic.com>

lezama committed:
not finished check

Co-Authored-By: Daniel Walmsley <goldsounds@gmail.com>

lezama committed:
use assertSame

Co-Authored-By: Daniel Walmsley <goldsounds@gmail.com>

kraftbj commented:
`branch-7.9` via c4858f022

kraftbj commented:
`sync` released via https://github.com/Automattic/jetpack/tree/automattic/jetpack-sync%401.3.2
